audio/out/ao_oss.c

```diff
@@ -0,0 +1,439 @@
+#endif
+    {-1, -1}
+};
+
+#define MP_WARN_IOCTL_ERR(__ao) \
+    MP_WARN((__ao), "%s: ioctl() fail, err = %i: %s\n", \
+        __FUNCTION__ ,errno, strerror(errno))
+
+
+static void uninit(struct ao *ao);
+static void reset(struct ao *ao);
+static bool set_pause(struct ao *ao, bool pause);
+
+
+static int format2oss(int format)
+{
+    for (size_t i = 0; format_table[i][0] != -1; i++) {
+        if (format_table[i][1] == format)
+            return format_table[i][0];
+    }
+    return -1;
+}
+
+static bool try_format(struct ao *ao, int *format)
+{
+    struct priv *p = ao->priv;
+    int oss_format;
+
+    oss_format = format2oss(*format);
+    if (oss_format == -1 && af_fmt_is_spdif(*format))
+        oss_format = AFMT_AC3;
+
+    if (oss_format == -1) {
+        MP_VERBOSE(ao, "Unknown/not supported internal format: %s\n",
+                   af_fmt_to_str(*format));
+        *format = 0;
+        return false;
+    }
+
+    return (ioctl(p->dsp_fd, SNDCTL_DSP_SETFMT, &oss_format) == 0);
+}
+
+static int init(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+    struct mp_chmap channels = ao->channels;
+    struct mp_chmap_sel sel;
+    audio_buf_info info;
+    size_t i;
+    int format, samplerate, c, nchannels, reqchannels;
+    int best_sample_formats[AF_FORMAT_COUNT + 1];
+    const char *device = PATH_DEV_DSP;
+
+    /* Opening device. */
+    if (ao->device)
+        device = ao->device;
+    MP_VERBOSE(ao, "using '%s' audio device\n", device);
+    p->dsp_fd = open(device, O_WRONLY);
+    if (p->dsp_fd < 0) {
+        MP_ERR(ao, "Can't open audio device %s: %s\n",
+               device, mp_strerror(errno));
+        goto fail;
+    }
+#if defined(FD_CLOEXEC) && defined(F_SETFD)
+    fcntl(p->dsp_fd, F_SETFD, FD_CLOEXEC);
+#endif
+
+    /* Selecting sound format. */
+    format = af_fmt_from_planar(ao->format);
+    af_get_best_sample_formats(format, best_sample_formats);
+    for (i = 0; best_sample_formats[i]; i++) {
+        format = best_sample_formats[i];
+        if (try_format(ao, &format))
+            break;
+    }
+    if (!format) {
+        MP_ERR(ao, "Can't set sample format.\n");
+        goto fail;
+    }
+    MP_VERBOSE(ao, "sample format: %s\n", af_fmt_to_str(format));
+
+    /* Channels count. */
+    if (af_fmt_is_spdif(format)) {
+        /* Probably could be fixed by setting number of channels;
+         * needs testing. */
+        if (channels.num != 2) {
+            MP_ERR(ao, "Format %s not implemented.\n", af_fmt_to_str(format));
+            goto fail;
+        }
+    } else {
+        memset(&sel, 0x00, sizeof(sel));
+        for (i = 0; i < MP_NUM_CHANNELS + 1; i++) {
+            mp_chmap_sel_add_map(&sel, &oss_layouts[i]);
+        }
+        if (!ao_chmap_sel_adjust(ao, &sel, &channels))
+            goto fail;
+        nchannels = reqchannels = channels.num;
+        /* We only use SNDCTL_DSP_CHANNELS for >2 channels, in case some
+         * drivers don't have it. */
```

Comment:
Which drivers don't have it? If they no longer exist, get rid of this special casing for stereo and with it the somewhat unpleasant `goto stereo` thing.

Reply:
This part from original code, I am not an author of this :)
Probably author mean that some audio devices can not support more than 2 channels.
Any way, docs say: http://manuals.opensound.com/developer/oss_error_codes.html
> SNDCTL_DSP_STEREO is an obsolete ioctl call that supports only mono (0) or stereo (1). For larger number of channels you need to use SNDCTL_DSP_CHANNELS instead. 

I will rewrite this code block.

Reply:
> This part from original code, I am not an author of this :)

You cannot say this, not even with a smiley. It's your PR, and you MUST understand why every line is there. It's fine to copy it from elsewhere (license permitting etc) but that doesn't give you a free pass to not fully understand the code which you submit.